### PR TITLE
feat: clean stale provisioner files

### LINKED ISSUE #7107: clean up coder cache
# Problem

We write a whole bunch of data to `CODER_CACHE_DIRECTORY` but we never seem to clean it out.
This can end up consuming the entire partition, potentially leading to service interruption / outage.

# Solution

Periodically clean `CODER_CACHE_DIRECTORY`.
- The simplest approach here is probably to remove files older than a configurable threshold.
- An alternative approach could be to remove files, starting with the oldest first, until the size of the cache directory is below a configurable threshold.
- Care will need to be taken that we do not perform this task while there are in-progress provisioner jobs, as this could cause them to fail.
  - We may need to provide a mechanism to temporarily "lock" local provisioners and pause them from accepting new jobs. We could potentially accomplish this by simply locking the jobs table.

### PR DESCRIPTION
**Pull request by mtojek**
Fixes: https://github.com/coder/coder/issues/7107

This PR introduces cleanup logic for stale provisioner files:
* unfinished, orphan provisioner sessions
* unused plugins based on the last "access time" attribute

_I tried to implement a really dumb sweeper that doesn't care about the content but wouldn't like to end up with partially removed plugin files, hence the implementation is plugin-aware._

TODO:
- [x] unit tests for `provisioner/terraform/cleanup.go`
- [x] unit tests for `provisionersdk/cleanup.go`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -197,6 +197,8 @@ require (
 	tailscale.com v1.46.1
 )
 
+require github.com/djherbis/times v1.5.0
```
Comment (johnstcn) [inline go.mod:200]: It's super annoying that this isn't exposed in the standard library.
I'm slightly concerned that this module doesn't appear to be very active, but it has no dependencies, so maybe that's OK?

File: provisioner/terraform/cleanup.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+// latestAccessTime walks recursively through the directory content, and locates
+// the last accessed file.
+func latestAccessTime(fs afero.Fs, pluginPath string) (time.Time, error) {
+	var latest time.Time
+	err := afero.Walk(fs, pluginPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		accessTime := info.ModTime() // fallback to modTime if accessTime is not available (afero)
```
Comment (johnstcn) [inline provisioner/terraform/cleanup.go:139]: Just be aware that modtime is not guaranteed (see: https://www.kernel.org/doc/html/latest/filesystems/api-summary.html?highlight=nocmtime#c.file_update_time)

Howver, I would _assume_ it to at least be the time the file was created.

**Review by johnstcn — APPROVED** [review]

File: go.mod (hunk 1 above)
Comment (mtojek) [reply go.mod:200]: I did the self-evaluation of the module, and I agree with you. Fortunately, this is just a basic syscall, so in theory, we could "port" it to coder.

File: provisioner/terraform/cleanup.go (hunk 2 above)
Comment (mtojek) [reply provisioner/terraform/cleanup.go:139]: Yes, that is correct. I was thinking about running a self-test on the cache directory, but on the other hand I wouldn't like to overcomplicate things.

**Comment by mtojek** [comment]
@johnstcn If you don't mind, I will merge this after the weekend to prevent any unexpected issues 👍

**Comment by johnstcn** [comment]
> @johnstcn If you don't mind, I will merge this after the weekend to prevent any unexpected issues 👍

👍 
```
chmod 444 /proc/sys/days/friday
```

**Pull request merged**